latest/index.bs

```diff
@@ -177,308 +177,99 @@ For this example we assume an image with 5 dimensions and axes called `t,c,z,y,x
-                |   |             # `uns` groups MAY contain groups, dataframes, dense arrays, and sparse arrays.
-                |   |
-                |   ├── .zgroup
-                |   |
-                |   ├── .zattrs   # `.zattrs` MUST contain `"encoding-type"`, which is set to `"csr_matrix"` by AnnData.           
-                |   │             # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.1.0"` by AnnData.
-                |   ...
-                |
-                ├── dataframe_0   # You MAY add dataframe group containers.
-                |   |                 # dataframes MAY be in the `uns` group or in a subgroup.
-                |   │
-                |   ├── .zgroup
-                |   │                     
-                |   ├── .zattrs       # `.zattrs` MUST contain `"_index"`, which is the name of the column in obs to be used as the index.           
-                |   │                 # `.zattrs` MUST contain `"column-order"`, which is a list of the order of the non-_index columns.
-                |   │                 # `.zattrs` MUST contain `"encoding-type"`, which is set to `"dataframe"` by AnnData.
-                |   │                 # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.2.0"` by AnnData.
-                |   │      
-                |   └── col_0         # Each column in the obs table is a 1D zarr array.
-                |       ├── .zarray   # Each columns MUST be chunked the same, but the chunking may be chosen by the user.
-                |       │
-                |       └─ 0
-                |
-                ├── dense_array       # You MAY dense arrays as n n-dimensional zarr arrays.
-                |   │                 # `dense_array` MUST not be a complex type (i.e., MUST be a single type)
-                |   │                 # `dense_array` MAY be chunked as the user desires.
-                |   |                 # `dense array` MAY be in the `uns` group or in a subgroup.
+                ├── .zattrs      # `.zattrs` MUST contain "column-order", which is a list of the order of the columns.
+                |                # `.zattrs` MUST contain "annotated-data", which is specified above.
+                |     
+                ├── column_1     # The table group MAY hold zero, one, or an arbitrary number of column arrays.
                 |   |
                 |   ├── .zarray
+                |   ├── .zattrs
+                |   |
                 |   ├── 0.0
-                |   │   ...
+                |   |   ...
                 |   └── n.m
                 |
-                └── sparse_array  # You MAY add sparse arrays as a zarr group for each sparse array.
-                    |             # sparse arrays MAY be in the `uns` group or in a subgroup.
-                    |
-                    ├── .zgroup
+                └── row_names    # The table group SHOULD contain a 1D array of strings of length n called `row_names`.
```

Comment:
It seems that `SHOULD` have `row_names` is a strong requirement. In most cases I can think of, each row will be e.g. a Cell ID and mostly these won't have names for each row, so I can't satisfy this requirement.
Perhaps `MAY` is better here?
Would tools that read this data be expected to handle this column differently from any another string columns e.g. called "cell_names" or "sample_names"? If I have a existing columns called "cell_names" and "sample_names" then it would only duplicate data if I have to add a "row_names" column too? So maybe this isn't really needed?

Reply:
From my point of view, annotating an axis of an array attributes some categorical meaning to the hyperslices obtained by slicing orthogonal to that axis. Therefore, it seemed natural to me to be able to refer to such a hyperslice by a name (or an ID). So natural in fact, that I would suggest having default names `Row 1, Row2, ...` in case this column is not present. In this regard, `MAY` seems to weak for me.

As for the redundancy, would it help to rename this to `names`, `IDs`, `identifier`, ... and allow Strings or integers as data type?

Reply:
SHOULD --> MAY is a schema-weakening change that is easy to roll out at any time. I think we should start with SHOULD and if people don't find row_names useful, we can change it to MAY later.

Allowing ints as well as strings seems reasonable to me.

I don't see why redundancy is an issue; row_names is not required. If there is a column that conceptually makes sense to think of as row names, you could just make that array row_names instead of a column. Presumably, the viewer will display the row names, so you wouldn't lose that column, it would just be moved to a more prominent place.

Reply:
Also, now that I am thinking about this more... I think we should require that IF row_names exists, it MUST be unique, i.e., cannot contain duplicates. Having unique row names guarantees you have no duplicate rows, so your query will not accidentally grab 2 rows when you were expecting 1. Unique row names are so natural that in R, dataframes always have them. (By default, 'rownames' is just the row numbers, as a vector of strings. R's 'rownames' can be strings or ints.) 

I think it's probably not the file format's responsibility to provide deafult row names--any software that's reading and working with tables can do that. Nevertheless, we should encourage the user to provide them if it makes sense for their dataset, since row names are a useful and commonly used data structure. Considering that R is designed for tabular data and is extremely popular in the genomics community, I think copying R's design choices is not a bad idea.

Reply:
To summarize this discussion: I [added](https://github.com/minnerbe/ngff/commit/75e8b0134240ba05e4ca8b93afc7bae7c245adc5) an attribute "row-names" that MAY be present and refers to a 1d-column of strings or integers that should be used as row names. What do you think about this?

Reply:
Excellent, but I would change how it's presented in the spec. I'll suggest some new text in a private conversation. As usual, I am concerned with story-telling. I think we should present the three new properties ("row-names", "column-order", and "annotated-data") together in the text, as well as in the tree diagram.

I would delete the row_names directory from the tree diagram. This visual design suggests that the directory should be named row_names, which is not the case. It also makes the spec look more complicated than it really is.